midterm-2023-11-07/code/5048-dijana-damcevska/scripts.ts

```diff
@@ -0,0 +1,225 @@
+interface Oscars {
+    award: string;
+}
+
+interface Cast {
+    actor: string;
+    character: string;
+}
+
+interface Movie {
+    id: number;
+    title: string;
+    year: number;
+    director: string;
+    genre: string[];
+    plot: string;
+    cast: Cast[];
+    oscars: Oscars[];
+    rating: number;
+}
+
+type MovieSorter = (first: Movie, second: Movie) => number;
+
+
+document.addEventListener("DOMContentLoaded", moviesCode)
+
+let movies: Movie[] = [];
+
+async function moviesCode() {
+    const data = await loadData();
+    movies = data;
+
+    fillGenres(movies);
+    displayMovies(movies);
+
+    const titleSort = document.getElementById("sort-title")!;
+    titleSort.addEventListener("click", sortByTitle);
+
+    const idSort = document.getElementById("sort-id")!;
+    idSort.addEventListener("click", sortById);
+
+    const yearSort = document.getElementById("sort-year")!;
+    yearSort.addEventListener("click", sortByYear);
+
+    const directorSort = document.getElementById("sort-director")!;
+    directorSort.addEventListener("click", sortByDirector);
+
+
+    const applyFilterButton = document.getElementById("apply-filter")!;
+    applyFilterButton.addEventListener("click", applyFilter);
+}
+
+const titleSorter: MovieSorter = (first, second) => first.title.localeCompare(second.title);
+const idSorter: MovieSorter = (first, second) => first.id - second.id;
+const directorSorter: MovieSorter = (first, second) => first.director.localeCompare(second.director);
+const yearSorter: MovieSorter = (first, second) => first.year - second.year;
+
+
+const sortByTitle = () => {
+    const sortedMovies = movies.toSorted(titleSorter);
+    displayMovies(sortedMovies);
+}
+
+const sortById = () => {
+    const sortedMovies = movies.toSorted(idSorter);
+    displayMovies(sortedMovies);
+}
+
+const sortByDirector = () => {
+    const sortedMovies = movies.toSorted(directorSorter);
+    displayMovies(sortedMovies);
+}
+
+const sortByYear = () => {
+    const sortedMovies = movies.toSorted(yearSorter);
+    displayMovies(sortedMovies);
+}
+
+
+const fillGenres = (movies: Movie[]) => {
+    const filter = document.getElementById("genre-filter")!
+    const genres = new Set<string>();
+
+    for (const movie of movies) {
+        const movieGenres = movie.genre;
+        for (const genre of movieGenres) {
+            genres.add(genre);
+        }
+    }
+
+    for (const genre of genres) {
+        const option = document.createElement("option");
+        option.value = genre;
+        option.innerHTML = genre;
+        filter.appendChild(option);
+    }
+}
+
+
+const applyFilter = () => {
+
+    const titleElement = document.getElementById("title-filter") as HTMLSelectElement;
```

Comment:
Well, not everything is a `HTMLSelectElement`, the `value` property still works. Getting filter values +8